sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+the queue's asynchronous handler.
+
+### No runtime exception for speculative compilation
+
+It is currently common for the runtime to speculatively compile some kernels.
+For example, DPC++ may bundle all kernels from the same translation unit
+together into a single device image.  When the application submits one kernel K
+to a device D, the runtime actually compiles all kernels in K's device image
+for device D.  Let's assume in this example that the kernel K uses only
+features that are supported by D.  It would be illegal for the runtime to throw
+an exception in such a case just because some other kernel in the same device
+image uses a feature that is not supported by device D.
+
+
+## Design to implement required diagnostics
+
+### Changes to DPC++ headers
+
+With the exception of the `[[sycl::reqd_work_group_size()]]` and
+`[[sycl::reqd_sub_group_size()]]` attributes, all kernel optional features
+are associated with some device aspect.  For example, the `sycl::half` type
+is an optional feature which is only supported on devices that have the
+`aspect::fp16` aspect.  We can therefore use device aspects as a way to
+describe the set of optional features that a kernel uses (with the exception
+of the required work-group or sub-group size).
+
+As will see later, it will be very helpful to decorate all APIs in DPC++
+headers that correspond to optional kernel features with the
+`[[sycl::requires()]]` attribute.  For example, the declaration of the
+`sycl::half` type would look like this:
+
+```
+using half [[sycl::requires(has(aspect::fp16))]] = cl::sycl::detail::half_impl::half;
+```
+
+In cases where the optional feature corresponds to use of a class (e.g.
+`sycl::atomic_ref`), the declaration can look like this:
+
+```
+template</*...*/>
+class [[sycl::requires(has(aspect::fp64))]] atomic_ref {
+  /* ... */
+};
+```
+
+(We can use partial specialization tricks to decorate `atomic_ref` with the
+attribute only when the underlying type is 64-bits.)
+
+In cases where the optional feature corresponds to a function, we can decorate
+the function's declaration with the attribute like so (demonstrating a
+hypothetical AMX multiplication extension):
+
+```
+[[sycl::requires(has(aspect::ext_intel_amx))]]
+void amx_multiply();
+```
+
+These attributes provide an explicit mapping between each optional kernel
+feature and its associated aspect.
+
+Unfortunately, the fundamental type `double` is also an optional kernel
+feature.  Since there is no type alias for `double`, there is no convenient
+place to add an attribute.  Instead, the front-end device compiler must behave
+as though there was an implicit `[[sycl::requires(has(aspect::fp64))]]`
+attribute for any device code that uses the `double` type.
+
+Note that the usage of `[[sycl::requires()]]` is slightly expanded here beyond
+the specified usage in the SYCL 2020 specification because we allow the
+attribute also on type alias declarations and class declarations.  If a device
```

Comment:
I think we should not update the 2020 spec to include this behavior because:

* It's not clear to me that this expanded behavior that I describe is useful for application developers.  I can see how it's useful for vendors to implement SYCL, but this is not a reason to expose it to applications as part of the supported API.

* We are reconsidering whether we want to rely so heavily on C++ attributes in future versions of the SYCL spec.  I'd rather not add more attributes to SYCL 2020.

* SYCL 2020 is already ratified, so we need a stronger motivation to add new features at this point.

Perhaps you are concerned about expanding the meaning of `[[sycl::requires()]]` without documenting the new behavior to users.  I also share this concern.  See my TODO below.  If we feel it is bad practice to expand the semantics of `[[sycl::requires()]]` beyond what we document to users, then we can instead introduce a new C++ attribute for this purpose, which we do NOT document to users.  For example, we could name it `[[intel::type_requires()]]` or `[[sycl_detail::type_requires()]]`.

Reply:
> I was not aware that end-users were not supposed to use this extension directly. Do you think there are no scenarios where it is needed?

Correct, I do not expect end-users to rely on this extended behavior of `[[sycl::requires()]]`, and I don't see a compelling use case right now.

Let's ask @erichkeane and @AaronBallman for their opinion about whether we should define a new (internal) attribute or extend the semantic of this existing attribute.  I would also like to get your review for this design in general, especially the parts that affect the compiler front-end.

Reply:
> Please note that this document is not an extension specification, but rather an implementation design

It references an unspecified feature which could be potentially user-visible (extending the `requires` attribute usage beyond specification), so I would generally see at as a problem, unless it is really not usable by end-users.

Reply:
After reading this conversation, I think it would be better to define a new internal attribute rather than expanding the definition of `sycl::requires`.  We already have a precedent for defining an internal (undocumented) attribute: `__attribute__((sycl_device))`, which is used in the expansion of `SYCL_EXTERNAL`.  I propose that the base name of this new attribute should be `type_requires`, but I'm open to the exact spelling:

* It could be `__attribute__((sycl_type_requires(aspect)))`, to match the existing `sycl_device` attribute.

* It could be `[[__sycl::type_requires(aspect)]]`, using the modern C++ attribute syntax and defining a new internal `__sycl::` attribute namespace.

Personally, I like the second option, but this is not the direction we chose for `sycl_device`.

One advantage to defining a new internal attribute for this purpose is to allow a "pedantic" mode that will warn the user of incorrect usage of the documented `[[sycl::requires()]]` attribute.

Reply:
+ 1 for second and __sycl - style. Maybe even `sycl::detail::type_requires`

Reply:
Unfortunately, C++ does not allow multiple levels of namespace attributes, so `sycl::detail::` is not possible.  If we use the modern C++ attribute syntax, we need to define a single attribute namespace for these internal attributes.

Reply:
I don't have any strong preference. 1 more variant - `sycl_detail::`.

Reply:
I also don't have a strong preference.  I suppose `sycl_detail` has a similar name to the regular C++ namespace `sycl::detail` which is reserved for implementation details, so I'm happy to go with that.

Whatever we choose for an attribute namespace here, we should use for all other internal DPC++ attributes going forward.

Reply:
How about a name like `__sycl_internal__`? I think that's more scary for users to write than `sycl_detail` and so they may be more likely to not try using it themselves.

Reply:
I have a slight preference for using the word "detail" rather than "internal", only because we've already set a precedent for "detail" with the internal namespace `sycl::detail`.

Out of curiosity, is there some rule (or precedent) for when you use no leading `__`, leading `__` AND trailing `__`, vs. just leading `__`?  For example, would it be better to name the attribute namespace `__sycl_detail` or `__sycl_detail__`?

Reply:
I'd be fine with `__sycl_detail__` as well, FWIW.

> Out of curiosity, is there some rule (or precedent) for when you use no leading __, leading __ AND trailing __, vs. just leading __? For example, would it be better to name the attribute namespace __sycl_detail or __sycl_detail__?

It's... a mess. :-D

The general rule is: code that exists as part of the implementation (C standard library, CXX standard library, intrinsic headers, SYCL, POSIX, etc) should be using reserved names for anything that's not part of the public API. This is done so that the implementation does not have to worry about user-defined macros causing the implementation headers to translate differently. That goes for all the identifiers, including attribute names, template parameters, function parameter names, private data members, everything in detail namespaces, etc.

For attributes specifically, to help implementers out, both GCC and Clang allow you to surround attributes with double square brackets. e.g., `__attribute__((__deprecated__("don't use this")))` and `__attribute__((deprecated("don't use this")))` are identical. When C++-style attributes were added, the same shenanigans were allowed there, but extended to also cover the vendor namespace. Thus, `[[clang::whatever]]`, `[[clang::__whatever__]]`, `[[__clang__::whatever]]`, and `[[__clang__::__whatever__]]` are also identical.

Unfortunately, sometimes these reserved names leak out into user code these days, like with `__has_cpp_attribute` or `__has_include`, which are perfectly fine for users to use because they're standardized features. But aside from some special cases like that, programmers typically understand that using something with a `__reserved` name means they're fiddling with something that doesn't belong to them and shouldn't be sad if their code breaks as a result.

Reply:
I'm planning to use the name `uses_aspects` for this internal attribute, so the full name would be `[[__sycl_detail__::uses_aspects()]]`.  However, given when you say above, I guess I should name it instead as `[[__sycl_detail__::__uses_aspects__()]]`.  Correct?

Reply:
Kinda probably not but maybe sorta who knows? ;-)

Clang *automatically* supports the double underscore bits, so I suppose you could name it `[[sycl_detail::uses_aspects()]]` in the FE, but from within the SYCL headers, you always use `[[__sycl_detail__::__uses_aspects__()]]`. We could also force the double underscores so no one ever forgets to use the protected name, and thus the FE would name it `[[__sycl_detail__::__uses_aspects__()]]` and it would be possible for someone to just slap a ridiculous number of underscores onto it as `[[____sycl_detail____::____uses_aspects____()]]` thanks to the automagical double underscore behavior.

I was thinking we'd have the FE name it `[[__sycl_detail__::uses_aspects()]]` just so that users can't use `sycl_detail` thinking it is a valid thing for them, but the library would use `[[__sycl_detail__::__uses_aspects__()]]`. But it would probably be fine to name it `[[sycl_detail::uses_aspects()]]` and just use the double underscores within the SYCL headers, too.

Reply:
I'd like this design document to use the same name that we will write in the DPC++ headers.  I think you are saying that this will be the form with double underscores on both the namespace and the attribute: `[[__sycl_detail__::__uses_aspects__()]]`.  Correct?

Reply:
Yes, that's correct, but may cause some confusion too. We want the DPC++ headers to use `[[__sycl__::__requires__()]]` for the same reason, but users are welcome to use `[[sycl::requires()]]` for their own code. I suspect we can use some prose to make sure this is sufficiently clear, though.

FWIW, the double underscore thing is more of a policy issue that's wider than this design doc. I filed https://github.com/intel/llvm/issues/3677 to track that as more of a general concern with the DPC++ headers.